### ISSUE TITLE
cutechess version check/update

### ISSUE THREAD

**Issue by vondele** [glinscott/fishtest#488]
If we have a specific version requirement on cutechess, we have no easy way to force an update of that on the worker side. One option would be to download the latest cutechess version from the books repo whenever there is a worker update, or this could be triggered by a flag in these worker updates. 

I'm not familiar enough with some of the infrastructure (i.e. windows portable installer), or cutechess/linux incompatibilities to suggest the right approach, I'm afraid.

**Comment by ppigazzini** [comment]
A very lazy solution (but with the advantage to be sure to don't use old files) is to add below here https://github.com/glinscott/fishtest/blob/3d116f5349da4aeeb76079df483908a0f9d4ec17/worker/updater.py#L43-L45 something like this:

```python
testing_dir = os.path.join(worker_dir, 'testing')
shutil.rmtree(testing_dir)
```

Moving the "testing" folder will save cutechess-cli custom builds (e.g. for OSX users) but it will pollute the worker folder, though.

```python
testing_dir = os.path.join(worker_dir, 'testing')
bkp_testing_dir = os.path.join(worker_dir, 'bkp_testing_dir')
shutil.move(testing_dir, bkp_testing_dir)
```

I have been testing for some weeks several solutions for the portable worker users:
 - the lazy one is to make a new archive with python 3 and up to date mingw-w64 and bat script, but I dislike to package *.exe (license, virus etc.)
 - the dreamed one is to write a powershell script to use [chocolatey](https://chocolatey.org/) to install msys2 and configure the worker (installing packages for gcc, python3 git etc.), but I dislike posh script language :)

**Comment by vondele** [comment]
I would use the moving solution for the directory... just in case users keep something valuable there, and even keep a history i.e. move to something like: mv testing testing.old.`date --iso-8601=seconds`

**Comment by ppigazzini** [comment]
@vondele I like more the idea to use a rotation scheme (e.g. .000 .001 .002 etc) to keep only max N (e.g. 3) old folders and to be able to add the folders names to .gitignore

**Comment by vondele** [comment]
even better, I agree.

**Comment by ppigazzini** [comment]
@vondele we could also move the sf binaries in the brand new "testing" folder and perhaps even the books...

**Comment by vondele** [comment]
at that point we can as well just do the move trick on the cutechess binary only...

**Comment by ppigazzini** [comment]
@vondele the new cutechess-cli archives contain only a binary and a text file for the license, the old ones shipped with several files. Anyway we do only a half dozen worker updates in a year, so we could simply rotate the folders to be sure to have a clean start state and keep the code simple.

**Comment by vondele** [comment]
yes, simplest is probably best.

**Comment by vondele** [comment]
BTW, moving directories...  custom_make.txt is not in the testing directory, and thus not a concern..

**Comment by ppigazzini** [comment]
Correct, "testing" is a disposable folder except for the rare case of a custom cutechess-cli build (I haven't an OSX system).

**Comment by ppigazzini** [comment]
@vondele we could also make fresh cutechess/cli builds merging this https://github.com/cutechess/cutechess/pull/551

**Comment by vondele** [comment]
Ideally the cutechess patch gets reviewed and merged before we adopt it, but it is not so clear what a timeline for that could be. However, I think a release of cutechess is something that traditionally takes a long term, so that's maybe not something we should wait for.

To fully benefit from that patch, we would also need worker and server changes (I think), that specify the start point in the book depending on the batch Id and take care of providing proper seeds. Once we make those changes, a cutechess upgrade would be mandatory.

**Comment by noobpwnftw** [comment]
It should calculate hashes of all existing files(cutechess, extracted book files) and match them with the server-side data, if anything isn't right delete the files and download new ones. At this moment we cannot know if a incomplete download may have produced an incomplete book on the worker.

**Comment by ppigazzini** [comment]
@noobpwnftw the new github repo has only zipped archives, if I'm not wrong should be possible to extract the files only w/ a complete download. Checking the hash should be done only in the update function, in order to keep simple the test of a new cutechess-cli build and to make possible to use a custom cutechess-cli for some exotic or old distro. Unfortunately cutechess-cli seems to have an useless version message:
```
$ ./cutechess-cli --version
cutechess-cli 1.1.0
Using Qt version 5.10.1

Copyright (C) 2008-2018 Ilari Pihlajisto and Arto Jonsson
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by ppigazzini** [comment]
@vondele I wrote the patch using your code of stockfish binaries, I will open the PR after fixing the worker update on Windows (never worked for me in these years).

**Comment by ppigazzini** [comment]
@vondele problem fixed, the updater copied all the fishtest files (worker and server, sic) in the parent folder, so with a custom folder name (like that used in the windows portable installer or in my worker test setup) the files were never updated.

**Comment by ppigazzini** [comment]
Opened #528

**Comment by OuaisBla** [comment]
I had a lot of trouble with this issue. The problems came on Wnidows when the book opening format was change from .pgn to .epd.

The worker stalls at the "Indexing opening suites" message. The local patch is to replace Cutechess-cli.exe from the web site and dll into the fishtest\worker directory.

To fix that, we should update the Fishtest-Worker-Portable.exe with the latest version of the cutechess-cli.exe. That is only a zip files. 

That's the easiest and most convenient way on Windows to run the worker. For now, all Windows user can't easily run the worker without the workaround.

**Comment by ppigazzini** [comment]
@OuaisBla the true problem is that "Fishtest-Worker-Portable.exe" contains "cutechess-cli".
On Windows I ran very easily the worker either with WSL or msys2:
https://github.com/glinscott/fishtest/wiki/Running-the-worker-on-Windows

**Comment by OuaisBla** [comment]
If you upgrade the cutechess-cli to the lastest version 1.0.0 or higher, it works.

The following instruction, on the wiki page you referenced, leads to a non working worker:

"If you are not a Linux user, better to use the following instructions to do a manual installation or use A portable fishtest worker for Windows 64 bit."

The Fishtest-Worker-Portable.exe should either be removed, or much better fixed, by including an upgrade version of cutechess-cli (like 1.0.0 or higher) that is working. I definitely recommand fixing it, since it is tghe easiest way on Windows to have a worker running. This is a very convenient and easy to setup. More workers, more tests, more fun.

**Comment by ppigazzini** [comment]
@OuaisBla yes, I know, I wrote that wiki page (and many others) :)

If there is not a "testing" folder the worker automatically make it and download the latest version of "cutechess-cli.exe", so adding "cutechess-cli.exe" in "Portable" was/is a big design error.

The update "Fishtest-Worker-Portable.exe" is in my ToDo but I don't like the licensing mess:
https://groups.google.com/d/msg/fishcooking/4_ausUwMXP0/fWuHft-sEgAJ

I routinely test all the worker setup scripts but something a bug can slip, so please let me know if you find one, thank you :)

**Comment by ppigazzini** [comment]
Closed by #528 and #785
